File: e2e_tests/tools/custom/python/identity_tool.py
#!/usr/bin/env python3
"""Identity tool using the Pekobot Tool SDK.

This tool demonstrates context injection by echoing back the injected identity.
"""

# Try to import from the installed SDK, fall back to local copy for development
try:
    from pekobot_tool import tool
except ImportError:
    # Fallback: use local adapter when SDK not installed
    import sys
    import os
    sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "..", "..", "..", "tools", "python", "pekobot_tool", "src"))
    from pekobot_tool import tool


@tool(
    name="identity_tool",
    description="Echo back injected identity parameters (for testing context injection)",
    parameters={
        "type": "object",
        "properties": {
            "message": {
                "type": "string",
                "description": "Optional message to echo back"
            }
        }
    },
    reserved=["session_id", "agent_id", "run_id", "workspace"]
)
def identity_tool(
    message: str = "",
    session_id: str = "",
    agent_id: str = "",
    run_id: str = "",
    workspace: str = "",
):
    """Echo back the injected identity parameters.
    
    This tool is useful for verifying that context injection is working correctly.
    The reserved parameters are injected by Pekobot at runtime.
    
    Args:
        message: Optional message to echo back
        session_id: Injected session ID (reserved)
        agent_id: Injected agent ID (reserved)
        run_id: Injected run ID (reserved)
        workspace: Injected workspace path (reserved)
    
    Returns:
        Dictionary with identity information
    """
    injection_working = bool(agent_id and session_id)
    
    return {
        "success": True,
        "message": message or "Context injection test",
        "injected_identity": {
            "agent_id": agent_id or "NOT_INJECTED",
            "session_id": session_id or "NOT_INJECTED",
            "run_id": run_id or "NOT_INJECTED",
            "workspace": workspace or "NOT_INJECTED",
        },
        "injection_working": injection_working,
        "verification": "Context injection is " + ("WORKING" if injection_working else "NOT WORKING"),
    }


if __name__ == "__main__":
    identity_tool.run()

File: e2e_tests/tools/custom/node/identity_tool.js
#!/usr/bin/env node
/**
 * Identity Echo Tool - E2E Test for Context Injection Verification (Node.js)
 * 
 * This tool explicitly reports back the injected identity parameters
 * to verify that context injection is working correctly.
 */

const { tool, run } = require('./pekobot_adapter');

const identityTool = tool(
    {
        name: "echo_identity",
        description: "Echo back the injected identity parameters to verify context injection",
        llm_description: `
Use this tool to verify that identity parameters (agent_id, session_id) are being injected.
The tool will return the injected values, which should be:
- agent_id: The name of the calling agent
- session_id: The current session ID
- run_id: The current run ID

Call this when asked to verify context injection is working.
`,
        parameters: {
            message: {
                type: "string",
                description: "Optional message to echo back"
            }
        },
        reserved: ["session_id", "agent_id", "run_id", "workspace"]
    },
    async ({ message, session_id, agent_id, run_id, workspace }) => {
        /**
         * Echo back identity parameters to verify context injection.
         * 
         * @param {string} message - Optional message from LLM
         * @param {string} session_id - Injected by Pekobot
         * @param {string} agent_id - Injected by Pekobot
         * @param {string} run_id - Injected by Pekobot
         * @param {string} workspace - Injected by Pekobot
         */
        
        // Check if parameters were actually injected
        const injectionWorking = (
            agent_id && agent_id !== "not_injected" &&
            session_id && session_id !== "not_injected"
        );
        
        return {
            success: true,
            message: message || "Context injection test",
            injected_identity: {
                agent_id: agent_id || "NOT_INJECTED",
                session_id: session_id || "NOT_INJECTED",
                run_id: run_id || "NOT_INJECTED",
                workspace: workspace || "NOT_INJECTED"
            },
            injection_working: injectionWorking,
            verification: "Context injection is " + (injectionWorking ? "WORKING" : "NOT WORKING")
        };
    }
);

run(identityTool);
